test/Tester/DependencyInjectionDisambiguationTests.cs

```diff
@@ -0,0 +1,81 @@
+﻿
+using System;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Xunit;
+using Orleans;
+using Orleans.Runtime.Configuration;
+using Orleans.TestingHost;
+using TestExtensions;
+using UnitTests.GrainInterfaces;
+
+namespace UnitTests.General
+{
+    [TestCategory("DI")]
+    public class DependencyInjectionDisambiguationTests : OrleansTestingBase, IClassFixture<DependencyInjectionDisambiguationTests.Fixture>
+    {
+        private readonly Fixture fixture;
+
+        public class Fixture : BaseTestClusterFixture
+        {
+            protected override TestCluster CreateTestCluster()
+            {
+                var options = new TestClusterOptions();
+                options.ClusterConfiguration.UseStartupType<MyStartup>();
+                return new TestCluster(options);
+            }
+        }
+
+        public DependencyInjectionDisambiguationTests(Fixture fixture)
+        {
+            this.fixture = fixture;
+        }
+
+        [Fact, TestCategory("BVT"), TestCategory("Functional")]
+        public async Task DisambiguateByKeyTest()
+        {
+            IDIDisambiguationGrain grain = this.fixture.GrainFactory.GetGrain<IDIDisambiguationGrain>(GetRandomGrainId());
+            int actual0 = await grain.GetDisambiguatedValue(0);
+            Assert.StrictEqual(0, actual0);
+            int actual1 = await grain.GetDisambiguatedValue(1);
+            Assert.StrictEqual(1, actual1);
+        }
+
+        [Fact, TestCategory("BVT"), TestCategory("Functional")]
+        public async Task DisambiguateByNameTest()
+        {
+            IDIDisambiguationGrain grain = this.fixture.GrainFactory.GetGrain<IDIDisambiguationGrain>(GetRandomGrainId());
+            string actualThis = await grain.GetDisambiguatedValue("this");
+            Assert.StrictEqual("this", actualThis);
+            string actualThat = await grain.GetDisambiguatedValue("that");
+            Assert.StrictEqual("that", actualThat);
+        }
+
+        private class SomeValue<TValue> : IValue<TValue>
+        {
+            public TValue Value { get; set; }
+        }
+
+        private class ValueServiceCollection<TValue> : IKeyedServiceCollection<TValue, IValue<TValue>>
+        {
+            public IValue<TValue> GetService(TValue name)
+            {
+                return new SomeValue<TValue> { Value = name };
```

Comment:
I understand this, as we've been talking a lot about it lately, but it probably makes more sense and it will be closer to our future usage if you pre-allocate a few keyed objects while constructing the ValueServiceCollection, and they this one just accesses them.
then you just register the singletons by passing an initialized collection